ee/tabby-webserver/src/repositories.rs

```diff
@@ -0,0 +1,113 @@
+use std::{collections::HashMap, path::PathBuf, str::FromStr};
+
+use anyhow::Result;
+use axum::{
+    body::boxed,
+    http::{header, Request, Uri},
+    response::{IntoResponse, Response},
+    Json,
+};
+use filenamify::filenamify;
+use hyper::Body;
+use mime_guess::mime;
+use serde::{Deserialize, Serialize};
+use tabby_common::SourceFile;
+use tokio::sync::OnceCell;
+use tower::ServiceExt;
+use tower_http::services::ServeDir;
+
+pub(crate) static DATASET: OnceCell<HashMap<DatasetKey, SourceFile>> = OnceCell::const_new();
+
+#[derive(Hash, PartialEq, Eq, Debug)]
+pub struct DatasetKey {
+    local_name: String,
+    rel_path: String,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct Repository {
+    name: String,
+    path: Option<String>,
+}
+
+impl Repository {
+    pub fn dataset_key(&self) -> DatasetKey {
+        DatasetKey {
+            local_name: self.name.clone(),
+            rel_path: self.path(),
+        }
+    }
+
+    pub fn name(&self) -> &str {
+        self.name.as_str()
+    }
+
+    pub fn path(&self) -> String {
+        self.path.clone().unwrap_or_default()
+    }
+}
+
+#[derive(Serialize)]
+struct ListDir {
+    entries: Vec<String>,
+}
+
+/// Load dataset
+pub async fn load_dataset() -> Result<()> {
+    let dataset = tokio::task::spawn_blocking(|| {
+        let all = SourceFile::all().unwrap();
+        let mut dataset = HashMap::new();
+        for mut file in all {
+            let key = DatasetKey {
+                local_name: filenamify(file.git_url.as_str()),
+                rel_path: file.filepath.clone(),
+            };
+            // exclude content from response data
+            file.content = "".to_string();
+            dataset.insert(key, file);
+        }
+        dataset
+    })
+    .await?;
+
+    DATASET.set(dataset)?;
+    Ok(())
+}
+
+/// Resolve a directory
+pub async fn resolve_dir(root: PathBuf, full_path: PathBuf) -> Result<Response> {
+    let mut read_dir = tokio::fs::read_dir(full_path).await?;
+    let mut entries = vec![];
+
+    while let Some(entry) = read_dir.next_entry().await? {
+        let path = entry.path();
+        let path = path.strip_prefix(&root)?;
+        let path = path.to_str().unwrap();
+        entries.push(path.to_string());
+    }
+
+    let body = Json(ListDir { entries }).into_response();
+    let resp = Response::builder()
+        .header(
+            header::CONTENT_TYPE,
+            mime::APPLICATION_OCTET_STREAM.as_ref(),
```

Comment:
For dir resolve, we set special content type `application/octet-stream` even though response is json

Reply:
Let's create vendor mime type: `application/vnd.directory+json`